content/posts/2025-01-15-perf-improvements.md

```diff
@@ -0,0 +1,235 @@
+
+**Fig. 4: eBPF agent kernel-space memory usage, on a single node**
+
+![Kernel-space memory]({page.image('perf-improvements-1-8/kernel-space-memory.png')})
+
+_NB: the "wide load" profile used here corresponds to 3000 workers rate-limited at 2 qps, using a more distributed pattern across pods/nodes/namespaces._
+
+### Overhead
+
+Some resource overhead is inevitable when observing traffic, but hopefully NetObserv makes it as low as possible. By lowering the CPU usage, we expect to lower the overhead: less CPU used by the eBPF agent means more CPU available for other tasks. When running low on CPU, it can make a difference.
+
+I used _hey_ again to measure the overhead in two different ways:
+
+- in terms of added latency
+- in terms of maximum throughput
+
+After stress-testing a target, _hey_ provides statistics such as latency percentiles, min/max/average, and the number of queries per second. It can be configured with rate limits, but for the purpose of this test I didn't use it. It's only limited by the number of workers, which is set to 50. I am also running a rate-limited _hey-ho_ in parallel, just to generate some background noise and warm NetObserv.
+
+Here is the output of the baseline run, without NetObserv:
+
+```
+Summary:
+  Total:	30.0017 secs
+  Slowest:	0.1207 secs
+  Fastest:	0.0006 secs
+  Average:	0.0015 secs
+  Requests/sec:	39863.3677
+  
+  Total data:	735520320 bytes
+  Size/request:	735 bytes
+
+Response time histogram:
+  0.001 [1]	|
+  0.013 [999261]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
+  0.025 [538]	|
+  0.037 [70]	|
+  0.049 [32]	|
+  0.061 [38]	|
+  0.073 [9]	|
+  0.085 [1]	|
+  0.097 [0]	|
+  0.109 [0]	|
+  0.121 [50]	|
+
+Latency distribution:
+  10% in 0.0008 secs
+  25% in 0.0009 secs
+  50% in 0.0011 secs
+  75% in 0.0014 secs
+  90% in 0.0018 secs
+  95% in 0.0023 secs
+  99% in 0.0038 secs
+
+Details (average, fastest, slowest):
+  DNS+dialup:	0.0000 secs, 0.0000 secs, 0.0044 secs
+  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0048 secs
+  req write:	0.0000 secs, 0.0000 secs, 0.0082 secs
+  resp wait:	0.0014 secs, 0.0006 secs, 0.1193 secs
+  resp read:	0.0001 secs, 0.0000 secs, 0.0223 secs
+
+Status code distribution:
+  [200]	1000000 responses
+```
+
+After that I ran the same command twice, first with NetObserv 1.7, second with NetObserv 1.8.
+
+In both cases, some values remained unchanged compared to the baseline: Fastest was always 0.0006s, Average 0.0015s, and same for the Fastest and Average details (DNS+dialup, DNS-lookup, req write, resp wait, resp read). These numbers are a good sign that NetObserv doesn't generate latency overhead for most of the requests. It's on higher percentiles that an overhead starts showing up.
```

Comment:
What is the percentile referring to?

Reply:
I'm giving more details below with the examples of p50 that shows no overhead and p99 having +0.7ms
So in 1.8, p99 is 0.0045s meaning that 99% of the responses arrived within 0.0045 seconds.